bkbit/scripts/gfftranslator.py

```diff
@@ -4,133 +4,346 @@
+            gene_annotations[stable_id] = curr_gene_annot
+    return gene_annotations.values()
+
+def resolve_ncbi_duplicates(stored_annotation, curr_annotation):
     """
-    taxon_name = {'9606': 'Homo sapiens', '10090': 'Mus musculus'}
-    with open(file=file, newline='') as csvfile:
-        reader = csv.DictReader(csvfile)
-        translatedannots = []
-        for row in reader:
-            myannotation = kbmodel.GeneAnnotation(id = row['gene_identifier_prefix'].upper().replace('ENE', 'ene') + ':' + row['gene_local_unique_identifier'],
-                                        symbol = row['name'],
-                                        name = row['name'],
-                                        description = row['description'],
-                                        #referenced_in = row['genome_annotation_label'],
-                                        #referenced_in = kbmodel.GenomeAnnotation(id = 'bican:assembly-' + row['genome_annotation_label']),
-                                        referenced_in = genomeAnnot,
-                                        molecular_type = row['gene_biotype'] if row['gene_biotype'] == 'protein_coding' else 'noncoding', #TODO: confirm if noncoding for the rest of the values or parse them as new enum values)
-                                        source_id = row['gene_local_unique_identifier'],
-                                        #in_taxon = [kbmodel.OrganismTaxon(id=taxon_uri[row['taxon_unique_identifier']])],
-                                        in_taxon = [orgTaxon],
-                                        in_taxon_label = taxon_name[row['taxon_unique_identifier']]
-                                        )
-            translatedannots.append(myannotation)
-    return kbmodel.AnnotationCollection(annotations=translatedannots)
+    Resolves duplicates between two GeneAnnotation objects with the same GeneID based on specific conditions.
+    Conditions: Among all duplicates with the same GeneID, keep one record per GeneID using the following priority:
+        1.	“description” is not NULL and “molecular_type” = ‘protein_coding’
+        2.	“description” is not NULL and “molecular_type” = ‘noncoding’
+        3.	“description” is NULL and “molecular_type” = ‘protein_coding’
+        4.	“description” is NULL and “molecular_type” = ‘noncoding
+
+    Parameters:
+    stored_annotation (GeneAnnotation): The stored gene annotation object.
+    curr_annotation (GeneAnnotation): The current gene annotation object.
+
+    Returns:
+    (GeneAnnotation): The gene annotation object with higher priority.
+
+    """
+    if stored_annotation.description is None and curr_annotation.description is not None:
+        return curr_annotation
+    elif stored_annotation.description is not None and curr_annotation.description is None:
+        return stored_annotation
+    elif stored_annotation.molecular_type == 'protein_coding' and curr_annotation.molecular_type == 'noncoding':
```

Comment:
above, you assign `molecular_type` to `kbmodel.BioType.protein_coding` and `kbmodel.BioType.noncoding`. You should check for those here (i.e. check for the kbmodel objects rather than the strings), just in case kbmodel ever changes what it does.

Reply:
fixed